File: qeli-shared/QeliShared/Geo/ProxyRoutePreset.cs
using System.Net;

namespace Qeli.Shared.Geo;

/// <summary>v2rayN-like routing presets for the local SOCKS/HTTP proxy.</summary>
public static class ProxyRoutePreset
{
    public const string ProxyAll = "proxy-all";           // RUv1-Всё / V4-全局
    public const string BypassRu = "bypass-ru";           // RUv1-Всё, кроме РФ
    public const string RuBlocked = "ru-blocked";         // RUv1-Заблокированное
    public const string BypassCn = "bypass-cn";           // V4-绕过大陆 (Whitelist)
    public const string GfwBlacklist = "gfw-blacklist";   // V4-黑名单

    public static IReadOnlyList<(string Id, string En, string Ru)> All { get; } =
    [
        (ProxyAll, "RUv1 / V4 — Everything (proxy all)", "RUv1-Всё / V4-全局"),
        (BypassRu, "RUv1 — Everything except RU", "RUv1-Всё, кроме РФ"),
        (RuBlocked, "RUv1 — Blocked only", "RUv1-Заблокированное"),
        (BypassCn, "V4 — Bypass mainland CN (whitelist)", "V4-绕过大陆 (Whitelist)"),
        (GfwBlacklist, "V4 — GFW blacklist", "V4-黑名单 (Blacklist)"),
    ];

    public static string Normalize(string? id) =>
        All.Any(p => p.Id == id) ? id! : ProxyAll;

    public static (string[] SiteTags, string[] IpTags) TagsFor(string presetId) =>
        Normalize(presetId) switch
        {
            BypassRu => (
                ["category-ru", "ru", "geolocation-ru", "private"],
                ["ru", "private"]),
            RuBlocked => (
                ["ru-blocked", "category-ru-blocked", "ru-blocked-community", "geolocation-!cn"],
                ["ru-blocked", "private"]),
            BypassCn => (
                ["cn", "geolocation-cn", "category-ads-all", "private"],
                ["cn", "private"]),
            GfwBlacklist => (
                ["gfw", "greatfire", "category-ads-all", "geolocation-!cn"],
                ["cn", "private"]),
            _ => ([], ["private"]),
        };

    /// <summary>
    /// True = send via tunnel proxy; false = dial direct (bypass VPN).
    /// Private IPs always direct.
    /// </summary>
    public static bool ShouldProxy(
        string presetId,
        string host,
        IPAddress? resolvedIp,
        GeoSiteIndex? site,
        GeoIpIndex? ip)
    {
        presetId = Normalize(presetId);
        if (resolvedIp != null && GeoIpIndex.IsPrivate(resolvedIp))
            return false;
        if (ip != null && resolvedIp != null && ip.MatchAny(resolvedIp, "private"))
            return false;

        return presetId switch
        {
            ProxyAll => true,

            BypassRu => !(
                (site != null && site.MatchAny(host, "category-ru", "ru", "geolocation-ru"))
                || (ip != null && resolvedIp != null && ip.MatchAny(resolvedIp, "ru"))),

            RuBlocked =>
                (site != null && site.MatchAny(host, "ru-blocked", "category-ru-blocked", "ru-blocked-community"))
                || (ip != null && resolvedIp != null && ip.MatchAny(resolvedIp, "ru-blocked")),

            BypassCn => !(
                (site != null && site.MatchAny(host, "cn", "geolocation-cn"))
                || (ip != null && resolvedIp != null && ip.MatchAny(resolvedIp, "cn"))),

            GfwBlacklist =>
                site != null && site.MatchAny(host, "gfw", "greatfire", "geolocation-!cn"),

            _ => true,
        };
    }
}

/// <summary>Process-wide proxy routing preference (set from desktop settings).</summary>
public static class ProxyRouteConfig
{
    public static string PresetId { get; set; } = ProxyRoutePreset.ProxyAll;
}

File: qeli-android/app/src/main/kotlin/com/qeli/geo/ProxyRoutePreset.kt
package com.qeli.geo

/** v2rayN-like routing presets (same IDs as Windows Qeli.Shared.Geo.ProxyRoutePreset). */
object ProxyRoutePreset {
    const val PROXY_ALL = "proxy-all"
    const val BYPASS_RU = "bypass-ru"
    const val RU_BLOCKED = "ru-blocked"
    const val BYPASS_CN = "bypass-cn"
    const val GFW_BLACKLIST = "gfw-blacklist"

    data class Entry(val id: String, val labelEn: String, val labelRu: String)

    val ALL = listOf(
        Entry(PROXY_ALL, "Proxy all", "Всё через VPN"),
        Entry(BYPASS_RU, "Bypass RU", "Всё, кроме РФ"),
        Entry(RU_BLOCKED, "RU blocked only", "Только заблокированное"),
        Entry(BYPASS_CN, "Bypass CN", "Обход материкового CN"),
        Entry(GFW_BLACKLIST, "GFW blacklist", "Чёрный список GFW"),
    )

    fun normalize(id: String?): String =
        if (ALL.any { it.id == id }) id!! else PROXY_ALL

    /** Site + IP tags to load from geosite.dat / geoip.dat for [presetId]. */
    fun tagsFor(presetId: String): Pair<List<String>, List<String>> =
        when (normalize(presetId)) {
            BYPASS_RU -> listOf("category-ru", "ru", "geolocation-ru", "su", "private") to
                listOf("ru", "private")
            RU_BLOCKED -> listOf("ru-blocked", "category-ru-blocked", "ru-blocked-community", "geolocation-!cn") to
                listOf("ru-blocked", "private")
            BYPASS_CN -> listOf("cn", "geolocation-cn", "category-ads-all", "private") to
                listOf("cn", "private")
            GFW_BLACKLIST -> listOf("gfw", "greatfire", "category-ads-all", "geolocation-!cn") to
                listOf("cn", "private")
            else -> emptyList<String>() to listOf("private")
        }

    /**
     * For TUN: CIDR codes from geoip that should be **excluded** from the tunnel
     * (direct / bypass) under the given preset. Empty = no geo excludes (proxy-all).
     */
    fun tunBypassIpCodes(presetId: String): List<String> =
        when (normalize(presetId)) {
            BYPASS_RU -> listOf("ru", "private")
            BYPASS_CN -> listOf("cn", "private")
            else -> emptyList()
        }

    /** geoip codes whose CIDRs are routed **into** the tunnel (proxy-only presets). */
    fun tunIncludeIpCodes(presetId: String): List<String> =
        when (normalize(presetId)) {
            RU_BLOCKED -> listOf("ru-blocked")
            else -> emptyList()
        }

    /** Bulk geoip excludes for full-tunnel bypass presets. */
    fun usesBypassExcludes(presetId: String): Boolean =
        tunBypassIpCodes(presetId).isNotEmpty()

    /** Split-tunnel: only listed destinations enter the VPN. */
    fun usesIncludeRoutes(presetId: String): Boolean =
        tunIncludeIpCodes(presetId).isNotEmpty() ||
            normalize(presetId) == RU_BLOCKED ||
            normalize(presetId) == GFW_BLACKLIST
}
